### ISSUE TITLE
Changes to forumdisplay_thread & forumdisplay_threadlist don't appear

### ISSUE THREAD

**Issue by BellsSparrow** [zingaburga/XThreads-MyBB-Plugin#32]
Using the newest github version of this plugin, MyBB 1.8.34 and PHP 8.2.3
None of the changes performed on forumdisplay_thread or forumdisplay_threadlist using a template prefix seem to get applied, but I'm getting a lot of this error:

`<error>
	<dateline>1687000271</dateline>
	<script>inc/plugins/xthreads.php</script>
	<line>365</line>
	<type>2</type>
	<friendly_type>Warning</friendly_type>
	<message>foreach() argument must be of type array|object, null given</message>
	<back_trace>#0  errorHandler->error() called at [/inc/class_error.php:153]
#1  errorHandler->error_callback() called at [/inc/plugins/xthreads.php:365]
#2  xthreads_tpl_cache() called at [/inc/plugins/xthreads.php(890) : eval()'d code:3]
#3  _objcont_1->cache() called at [/global.php:480]
#4  require_once() called at [/forumdisplay.php:24]
</back_trace>
</error>`

Could that have something to do with it?

**Comment by zingaburga** [comment]
Can you check whether the cache/xthreads_evalcache.php exists and that the cache/ directory is writable?

You can try editing a forum (no changes required, just edit + save) to force the cache to be rebuilt, if it's missing or out of date.

**Comment by BellsSparrow** [comment]
xthreads_evalcache.php exists, last updated around five days ago when I last tried to implement any change, and the directory is writable as well.

**Comment by zingaburga** [comment]
It's probably a good idea to force a cache rebuild, as described above, to rule out the possibility of corruption.

If that doesn't make a difference, in xthreads_evalcache.php, find the `xthreads_evalcacheForums` function.  Below that, you should a `switch($fid) {` line, followed by some `case [fid]:` lines (where [fid] is a forum ID).  
For the forum where the error is occurring, get the forum ID, match it with the relevant 'case statement', and see if it's returning a 'tplprefix' array.

It should look something like:

				case 2: return array(
					'tplprefix' => array_map('trim', explode(',', "prefix_")),
					// ... other lines here
				);

**Comment by BellsSparrow** [comment]
Forced the cache rebuild, which did not appear to make any difference. 

Regarding the array:
```
case 1: return array(
	'tplprefix' => array_map('trim', explode(',', "offplay_")),
	// other lines
	);
case 2: return array(
	'tplprefix' => array_map('trim', explode(',', "offplay_")),
	// other lines
);
```
Which matches the forum and subforum with those ids having that prefix.

**Comment by zingaburga** [comment]
That looks good, thanks.

Unfortunately, I'm unable to reproduce your issue or determine what it is 
It could be something else causing interference.

Are you able to reproduce this on a clean forum install?  If so, list the steps so I can reproduce and find the issue.

**Comment by BellsSparrow** [comment]
Sooo, I could not reproduce it on a clean forum install and instead ran into a different problem:
1. Removed the prior installation & uploaded all the files for MyBB 1.8.34
2. Followed all the installation steps, meeting all needed requirements, PHP Version 8.2.3
3. Only change applied to the settings is activating the logging of errors in error.log
4. Downloaded the newest github version of xThread, uploaded all files from the Upload folder to my directory
5. Upon Install & Activate of the Plugin, I get both a "has been activated sucessfully" as well as "_The XThreads settings file (cache/xthreads.php) is missing or corrupted! If you have modified this file, please revert the changes and try again, or if you have a backup copy of this file, please upload it. Otherwise, it is possible to try to regenerate the file if the version installed is v1.68 - if this is the case, please [click here] to try to regenerate the file._" This is the first time I have seen this message on the now four times I installed the plugin on a forum.
6. Clicked on the provided "Click Here" Link, get "XThreads settings file successfully rewritten - version set to v1.68 "
7. Went to edit the automatically created subforum with fid2, added "offplay_" as template prefix, got a whitescreen upon saving. The same happens when I try to create a new forum altogether.
Resulting error.log: [error.log](https://github.com/zingaburga/XThreads-MyBB-Plugin/files/11833377/error.log)

**Comment by zingaburga** [comment]
Thanks for trying all that.  Unfortunately I'm unable to reproduce, following your steps (the only difference being that I've got PHP 8.2.4 instead of 8.2.3, though I don't see that making a difference).

The issue you got at step 5 seems to suggest that the cache didn't initially write properly (though did when retried?!), and the errors in the log also relate to the forum cache lacking data it should have.  I don't know what it is, but there's something going on with MyBB's cache on your setup.

Your steps don't mention that you ever visited the non-admin section of the forum, but I'm guessing that happened at some point, because all the errors in your log point to code that isn't run in the admin panel.  This also means that the whitescreens you're getting upon saving, likely aren't causing the errors in the log.

Some things you can try:

* In AdminCP, Tools & Maintenance -> Cache Manager -> Rebuild Cache for 'forums'
* also click into 'forums' and see if "[xthreads_"* items are listed
* if the above is missing, check the forums table in the database to see if "xthreads_"* columns exist
* either enable the PHP `display_errors` setting (check php.ini) or find the value of PHP's `error_log` option (again, in php.ini).  Trigger the whitescreen again.  If you enabled error display, it should print out the error that caused the whitescreen, or if you have logging enabled (PHP logging, not MyBB error logging), an entry should be printed to the log

**Comment by BellsSparrow** [comment]
Thank you so much for helping me with this cache problem, which most likely had zero to do with your plugin. The forums table did not have the xthreads_ columns and continued to not have them even after I rebuild the Cache - both for 'forums' as well as simply... everything. In the end I simply re-installed the plugin again, the issue at step 5 didn't happen again and I still have no clue why it happened in the first place, but eh. Did learn how to activate the PHP error log, though.

Back to the original problem then. Successfully added a template prefix, created a new template for forumdisplay_threadlist with the prefix, simply added a "Test" to the legend aaaand - everything worked. It showed up in the forum with the prefix and didn't show up in a forum without prefix I created to check.
Next step: Slowly adding back in the other plugins I had installed. Starting with PHP in Templates (Version 2.2 from March 2021), since I remembered an old version of that already being on the "Known conflicting plugins" list. Bingo. The "Test" no longer appeared (but reappeared after deactivating PHP in Templates) and the error mentioned in the first comment showed up in my error log.
